### PR TITLE
jobs: Add DeployCollectionID key to flatpakref

### PR DESCRIPTION
**Pull request by mwleeds**
In preparation for Flathub deploying collection IDs to users, add the
DeployCollectionID key to generated flatpakref files. Although only
Flatpak >= 1.0.6 will honor it, older versions will ignore it without
producing an error.

The addition of DeployCollectionID to the flatpakrepo file and
ostree.deploy-collection-id to the repo metadata will have to be made
manually.

### CODE REVIEW

**Comment by mwleeds** [comment]
I don't have the tooling set up to see if this even compiles.

**Comment by alexlarsson** [comment]
Hmm, i think this has to be a per-repo config option whether to enable this. Also, I think we *never* should enable it for the build-repos (i.e. in commit, vs publish), as for it the collection id is that of the final repo.

Also, I thing leaving it out rather than empty is better for being disabled.

**Comment by mwleeds** [comment]
> Hmm, i think this has to be a per-repo config option whether to enable this. Also, I think we _never_ should enable it for the build-repos (i.e. in commit, vs publish), as for it the collection id is that of the final repo.

Okay, I don't have enough understanding of what the different repos do and why this would be bad for some of them. Could you either point to a similar repo config option (is it a different config than the ostree one?) or make the change yourself? That would probably be faster.

> Also, I thing leaving it out rather than empty is better for being disabled.

Sure

**Comment by alexlarsson** [comment]
> > Hmm, i think this has to be a per-repo config option whether to enable this. Also, I think we _never_ should enable it for the build-repos (i.e. in commit, vs publish), as for it the collection id is that of the final repo.
> 
> Okay, I don't have enough understanding of what the different repos do and why this would be bad for some of them. Could you either point to a similar repo config option (is it a different config than the ostree one?) or make the change yourself? That would probably be faster.

See the RepoConfig struct in app.rs, this is where you define the set of repos that flat-manager should update. For example, that is where you set the path to it, what collection id to use, and what gpg key to sign it with. What we want is an extra boolean here on whether to add a DeployCollectionID to it or not. Basically, we can add the code for this and deploy today, and then the day we want to flip it on for flathub we just change the config file.

But, there is also the build-repos. Like, each time we finish building an app we create a transient repository with just that app, just that version. This is used for testing a build (either a test one like from a PR, or a real build before you import it). We still generate a flatpakref for these though, and when generate_flatpakref() is called for the, the RepoConfig comes from the repo that is the final destination rather than the transient build repo, however build_id is set. 

So, basically, add a deploy_collection_id to RepoConfig. Then add the extra line to the flatpakref if RepoConfig.deploy_collection_id && RepoConfig.collection_id != None && maybe_build_id == None.

**Comment by mwleeds** [comment]
This should be good now

File: src/jobs.rs [hunk 1]
```diff
@@ -93,13 +93,25 @@ fn generate_flatpakref(ref_name: &String,
...
 
+    /* We only want to deploy the collection ID if the flatpakref is being generated for the main
+     * repo not a build repo.
+     */
+    let deploy_collection_id_config = match &repoconfig.collection_id {
+        Some(collection_id) => if repoconfig.deploy_collection_id && maybe_build_id == None {
+            format!("DeployCollectionID={}\n", collection_id)
+        } else {
+            "".to_string()
+        }
```
Comment (alexlarsson) [inline src/jobs.rs]: Shouldn't there be a comma here?

File: src/app.rs [hunk 2]
```diff
@@ -147,6 +147,7 @@ pub struct RepoConfig {
     pub suggested_repo_name: Option<String>,
     pub path: PathBuf,
     pub collection_id: Option<String>,
+    pub deploy_collection_id: bool,
```
Comment (alexlarsson) [inline src/app.rs]: I think you want  #[serde(default)] there too

Comment (mwleeds) [reply src/app.rs]: Makes sense, so it will default to false

File: src/jobs.rs (hunk 1 above)
Comment (mwleeds) [reply src/jobs.rs]: Yes, but I also noticed the pattern a few lines below of adding optional keys which seems simpler so I changed `DeployCollectionID` to work that way.
